counsel.el

```diff
@@ -4655,12 +4655,32 @@ selected color."
             :caller 'counsel-rhythmbox))
 
 ;;** `counsel-linux-app'
+(defalias 'counsel--xdg-data-home
+  (if (and (require 'xdg nil t)
+           (fboundp 'xdg-data-home))
+      #'xdg-data-home
+    (lambda ()
+      (let ((directory (getenv "XDG_DATA_HOME")))
+        (if (and directory (not (string= directory "")))
+            directory
+          "~/.local/share"))))
+  "Compatibility shim for `xdg-data-home'.")
+
+(defalias 'counsel--xdg-data-dirs
+  (if (and (require 'xdg nil t)
```

Comment:
Either the call to `require` can be extracted to top-level:

```el
(require 'xdg nil t)

(defalias 'counsel--xdg-data-home
  (if (fboundp 'xdg-data-home) ...))

(defalias 'counsel--xdg-data-dirs
  (if (fboundp 'xdg-data-dirs) ...))
```

or the two shims can be combined:

```diff
diff --git a/counsel.el b/counsel.el
index 559d34c..0d701fb 100644
--- a/counsel.el
+++ b/counsel.el
@@ -4655,32 +4655,29 @@ counsel-rhythmbox
             :caller 'counsel-rhythmbox))
 
 ;;** `counsel-linux-app'
-(defalias 'counsel--xdg-data-home
-  (if (and (require 'xdg nil t)
-           (fboundp 'xdg-data-home))
-      #'xdg-data-home
-    (lambda ()
-      (let ((directory (getenv "XDG_DATA_HOME")))
-        (if (and directory (not (string= directory "")))
-            directory
-          "~/.local/share"))))
-  "Compatibility shim for `xdg-data-home'.")
+(declare-function xdg-data-home "xdg")
+(declare-function xdg-data-dirs "xdg")
 
-(defalias 'counsel--xdg-data-dirs
-  (if (and (require 'xdg nil t)
-           (fboundp 'xdg-data-dirs))
-      #'xdg-data-dirs
+(defalias 'counsel--xdg-data-path
+  (if (require 'xdg nil t)
+      (lambda ()
+        (cons (xdg-data-home)
+              (xdg-data-dirs)))
     (lambda ()
-      (let ((path (getenv "XDG_DATA_DIRS")))
-        (if (and path (not (string= path "")))
-            (parse-colon-path path)
-          '("/usr/local/share/" "/usr/share/")))))
-  "Compatibility shim for `xdg-data-dirs'.")
+      (let ((home (getenv "XDG_DATA_HOME"))
+            (dirs (getenv "XDG_DATA_DIRS")))
+        (cons (if (and home (not (string= home "")))
+                  home
+                "~/.local/share")
+              (if (and dirs (not (string= dirs "")))
+                  (parse-colon-path dirs)
+                '("/usr/local/share/" "/usr/share/"))))))
+  "Return the XDG data directory search path as a list.
+This includes the base directory for user-specific data files.")
 
 (defcustom counsel-linux-apps-directories
   (mapcar (lambda (dir) (expand-file-name "applications" dir))
-          (cons (counsel--xdg-data-home)
-                (counsel--xdg-data-dirs)))
+          (counsel--xdg-data-path))
   "Directories in which to search for applications (.desktop files)."
   :group 'ivy
   :type '(repeat directory))
```

Reply:
I went with the first approach, and took the function definitions from `xdg.el`